tests/lib/linter/node-event-generator.js

```diff
@@ -187,6 +187,24 @@ describe("NodeEventGenerator", () => {
             ]
         );
 
+        assertEmissions(
+            "foo",
+            ["*:has(ExpressionStatement)"],
+            ast => [
+                ["*:has(ExpressionStatement)", ast], // Program
+                ["*:has(ExpressionStatement)", ast.body[0]] // ExpressionStatement
```

Comment:
I'm not sure if this is the intended behavior. Let's wait to see if this could be fixed in `esquery` if it turns out that it is a bug.

 https://github.com/estools/esquery/issues/122

Reply:
Let's wait 👍

Reply:
Given that there’s no response on the issue, how can we proceed here? Can we merge?

Reply:
/cc @mdjermanovic what do you think?

Reply:
I'm not sure how to proceed, there's still no response on estools/esquery#122. If this isn't intended behavior, maybe we shouldn't be promoting this feature yet.

This esquery feature is intended to match [CSS `:has()`](https://drafts.csswg.org/selectors-4/#has-pseudo), but that is an optional CSS feature and it seems that [none of the browsers support it](https://caniuse.com/css-has)  due to [performance concerns](https://www.chromestatus.com/feature/5794378545102848), so there's nowhere to check how that really works.

My interpretation of the spec text is that in the following example `ArrayExpression:has(ArrayExpression)` shouldn't match the inner ArrayExpression because `ArrayExpression ArrayExpression` doesn't match any of its child elements.

```js
/* eslint no-restricted-syntax: ["error",
    "ArrayExpression ArrayExpression", 
    "ArrayExpression:has(ArrayExpression)"   
]*/

[
  []
]
```

But it does, as in [this online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLFxuICAgIFwiQXJyYXlFeHByZXNzaW9uIEFycmF5RXhwcmVzc2lvblwiLCBcbiAgICBcIkFycmF5RXhwcmVzc2lvbjpoYXMoQXJyYXlFeHByZXNzaW9uKVwiICAgXG5dKi9cblxuW1xuICBbXVxuXSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2V0dGVyLXJldHVybiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19). Maybe it is the correct behavior, but I'm not sure where could we verify it.

Reply:
If browsers don’t support it, I think that’s a good argument that ESLint doesn’t need to support it.

Reply:
It should generally be avoided in rules, for performance reasons, as there may be a better way to write the rule.

It can be useful for user-defined selectors, like those in `no-restricted-syntax`, where all the user can do is to write a selector and there may be no alternative with the same functionality.

Reply:
I don't think that's a strong enough use case to be worth the effort for this feature. `no-restricted-syntax` is a convenience, and for anything more complicated, I'm fine with telling people to write a custom rule.

I'd say let's close this.

Reply:
I agree with closing. I already had concerns that this selector might be overused in rules, and now when it looks like it doesn't work as expected, it seems better that we don't include it.

It already can be used in ESLint (only the specificity calculations are wrong), but we don't need to officially promote it.

Reply:
I agree with closing as well 👍